src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractWorkspacePullDiagnosticsHandler.cs

```diff
@@ -76,23 +81,50 @@ private void OnLspTextChanged(object? sender, EventArgs e)
 
     private void UpdateLspChanged()
     {
-        Interlocked.Exchange(ref _lspChanged, 1);
+        lock (_gate)
+        {
+            // Loop through our map of source -> has changed and mark them as all having changed.
+            foreach (var category in _categoryToLspChanged.Keys.ToImmutableArray())
+            {
+                _categoryToLspChanged[category] = true;
+            }
+        }
     }
 
-    protected override async Task WaitForChangesAsync(RequestContext context, CancellationToken cancellationToken)
+    protected override async Task WaitForChangesAsync(string? category, RequestContext context, CancellationToken cancellationToken)
     {
+        // A null category counts a separate category and should track changes independently of other categories, so we'll add an empty entry in our map for it.
+        category ??= string.Empty;
+
         // Spin waiting until our LSP change flag has been set.  When the flag is set (meaning LSP has changed),
         // we reset the flag to false and exit out of the loop allowing the request to close.
         // The client will automatically trigger a new request as soon as we close it, bringing us up to date on diagnostics.
-        while (Interlocked.CompareExchange(ref _lspChanged, value: 0, comparand: 1) == 0)
+        while (!HasChanged())
```

Comment:
The race was here - when there are multiple diagnostic categories (and therefore multiple requests from the client for ws diagnostics), we would previously fail to close most of the requests when we saw an LSP change.  

There was 1 call to WaitForChangesAsync per request - however the update to the _lspChanged variable would be reset back to 0 as soon as only only one of the requests closed, leaving the rest still open.

This fixes the issue by storing the changed state per source, so the request for each source can separately observe the state change.